### ISSUE TITLE
Исследоваание семантического анализа

### ISSUE THREAD

**Issue by Stepan5024** [Stepan5024/Find-My-Doc#9]
# Семантическая близость текстов: Определение неизменности оригинального текста

## Суть
Разработать систему для семантической оценки близости двух текстов с целью проверки, был ли оригинальный текст изменен. Изменения могут включать замену слов на синонимы, перестановку слов, удаление или добавление частей текста. Система должна эффективно определять степень изменений и подтверждать неизменность оригинала.

## Описание
В данной задаче необходимо создать пайплайн, который позволяет:

1. **Векторизовать** тексты для их семантического сравнения.
2. **Сравнивать** новые тексты с базой оригинальных документов.
3. **Оптимизировать** процесс сравнения для обеспечения высокой скорости и точности.

### Функция 1: Векторизация и индексирование документов
- **Выбор модели для векторизации**:
  - Использовать современные модели трансформеров, специализированные для русского языка (например, `DeepPavlov/rubert-base-cased` или `sberbank-ai/ruBERT`).
- **Преобразование документов в векторное пространство**:
  - Применить энкодер трансформера для получения эмбеддингов каждого документа.
- **Создание индекса**:
  - Использовать библиотеки для поиска по векторному пространству (например, FAISS) для эффективного поиска ближайших соседей.
- **Хранение эмбеддингов**:
  - Сохранить эмбеддинги в базе данных или специализированном хранилище для быстрого доступа.

### Функция 2: Сравнение новых текстов с базой оригиналов
- **Обработка нового текста**:
  - Преобразовать новый текст в векторное представление с помощью выбранного энкодера.
- **Поиск ближайших соседей**:
  - Использовать индекс FAISS для поиска наиболее схожих оригинальных документов.
- **Точный алгоритм сравнения**:
  - Применить дополнительные методы сравнения (например, косинусное сходство, евклидово расстояние) для подтверждения степени близости.
- **Ограничение области поиска**:
  - Сначала сузить область поиска с помощью кластеризации или других методов, чтобы уменьшить количество сравнений и повысить скорость.

## Пример для тестирования

1. **Результат сравнения**:
   - Степень семантической близости между новым текстом и оригинальными документами, например:
     ```
     "Текст близок к оригиналу с идентификатором 123 (сходство: 0.95)"
     ```
   - Указание, был ли текст изменен:
     ```
     "Текст не был изменен."
     ```
     или
     ```
     "Текст был изменен (замена слов на синонимы, перестановка слов)."
     ```

## Инструкции по выполнению

1. **Выбор и настройка модели для векторизации**
   - Исследовать и выбрать SOTA модели для русского языка, например:
     - [DeepPavlov/rubert-base-cased](https://huggingface.co/DeepPavlov/rubert-base-cased)
     - [sberbank-ai/ruBERT](https://huggingface.co/sberbank-ai/ruBERT)
   - Установить необходимые библиотеки:
     ```bash
     pip install transformers faiss-cpu
     ```

2. **Векторизация базы документов**
   - Загрузить и подготовить базу документов.
   - Применить выбранный энкодер для преобразования каждого документа в векторное представление.
   - Создать и сохранить индекс FAISS для быстрого поиска.

   ```python
   from transformers import AutoTokenizer, AutoModel
   import torch
   import faiss
   import numpy as np

   # Загрузка модели и токенизатора
   tokenizer = AutoTokenizer.from_pretrained("DeepPavlov/rubert-base-cased")
   model = AutoModel.from_pretrained("DeepPavlov/rubert-base-cased")

   def encode(text):
       inputs = tokenizer(text, return_tensors="pt", truncation=True, padding=True)
       with torch.no_grad():
           outputs = model(**inputs)
       return outputs.last_hidden_state.mean(dim=1).squeeze().numpy()

   # Векторизация документов
   documents = [...]  # Список оригинальных документов
   vectors = np.array([encode(doc) for doc in documents]).astype('float32')

   # Создание индекса FAISS
   index = faiss.IndexFlatL2(vectors.shape[1])
   index.add(vectors)

   # Сохранение индекса
   faiss.write_index(index, "faiss_index.bin")

## Сравнение нового текста с базой

Векторизовать новый текст.
Найти ближайшие векторы в индексе FAISS.
Применить точный алгоритм сравнения для подтверждения степени близости.

Полезные материалы
Модели и библиотеки:
- Hugging Face Transformers
 - [FAISS (Facebook AI Similarity Search)](https://faiss.ai/)
- Sentence-BERT для русских текстов

Документация и учебные материалы:
- Руководство по FAISS
 - Обучение моделей на Hugging Face
 - Семантический поиск с использованием трансформеров
Примеры и туториалы:
1) Semantic Textual Similarity with BERT
2) Использование FAISS для поиска по векторному пространству
3) [Построение системы проверки текстов на плагиат](https://github.com/antontb/plagiarism-detector)
